feat: Add CGAN for timeseries

jfsantos-ds opened the pull request:
Closes #93

Commit:
tscwgan mvp

fabclmnt commented on src/ydata_synthetic/synthesizers/gan.py:
```diff
@@ -44,6 +44,8 @@ def __init__(
         self.noise_dim = model_parameters.noise_dim
         self.data_dim = model_parameters.n_cols
         self.layers_dim = model_parameters.layers_dim
+        self.cond_dim = model_parameters.condition
```
This parameter does not belong to the base model. 

Please added it only in the class model where this attribute is required.

Commit:
pr review

jfsantos-ds commented on src/ydata_synthetic/postprocessing/regular/inverse_preprocesser.py:
```diff
@@ -1,45 +1,46 @@
 # Inverts all preprocessing pipelines provided in the preprocessing examples
```
Changes in this script consist in extending the inverse support to the MinMaxScaler

jfsantos-ds commented on src/ydata_synthetic/preprocessing/timeseries/stock.py:
```diff
@@ -2,17 +2,30 @@
     Get the stock data from Yahoo finance data
```
Allowing subsetting of columns

jfsantos-ds commented on src/ydata_synthetic/preprocessing/timeseries/utils.py:
```diff
@@ -4,7 +4,7 @@
 import numpy as np
```
Enabling inverse_transform of the timeseries scaler (processor needs to be returned here)

jfsantos-ds commented on src/ydata_synthetic/synthesizers/regular/cramergan/model.py:
```diff
@@ -171,7 +171,7 @@ def save(self, path):
         super().save(path)
 
 
```
Noticed that the model was already imported in this script.
Just cleans up code, no changes come from it.

fabclmnt replied:
Please do not include changes in other models in the same PR as TSCGAN. 

Have this moved into a different branch.

Commit:
add example, remove added attribute of basemodel

remove changes on gitignore

removed unused n_feats argument

fabclmnt commented on src/ydata_synthetic/postprocessing/timeseries/inverse_preprocesser.py:
```diff
@@ -0,0 +1,17 @@
+
+from pandas import DataFrame
+
+def inverse_transform(data: List, processor: Union[Pipeline, ColumnTransformer, PowerTransformer, OneHotEncoder,
```
Why is the data of type list?

jfsantos-ds replied:
The preprocessing that was leveraged from the processed stock was returning a list of rolled sequences as numpy arrays.

Commit:
apply revisions + add typeguard

apply revisions

fabclmnt commented on src/ydata_synthetic/synthesizers/timeseries/tscwgan/model.py:
```diff
@@ -0,0 +1,248 @@
+                gen_input = concat([cond_seq, next(z_dist)], axis=1)
+                records = make_ndarray(make_tensor_proto(self.generator(gen_input, training=False)))
+                data.append(records)
+        return DataFrame(concat(data, axis=0))
```
@jfsantos-ds
Why are we concatenating the samples? 
And why are we assuming several conditions?

jfsantos-ds replied:
Changed, if accepting multiple condition sets the samples from different conditions shouldn't be concatenated.
Reverting to one array for a single condition set input.

fabclmnt reviewed (changes requested):
Revise sample method.

Commit:
integrate TSDataProcessor, revise sample method

Auto regressive timeseries sampling method

revert TS data processor integration

Commit:
flow fixes

jfsantos-ds commented on src/ydata_synthetic/synthesizers/timeseries/tscwgan/model.py:
```diff
@@ -0,0 +1,257 @@
+                data_.append(records)
+            data_ = hstack(data_)[:, :seq_len]
+            data.append(data_)
+        return array(vstack(data))
```
Regarding to a revision of this sample method I think:

1. We can force the number of samples to be the same as requested, i.e. truncate the last batch if needed
2. Condition converted to tensor could be taken outside the for loop
3. It does not seem to be truly auto-regressive since only the condition part of the produced records is being used as condition, we can add this capacity

Do you agree/have anything to add @fabclmnt ?
